### PR TITLE
Refine parallel unittest and enable it for CI

### PR DESCRIPTION
**Pull request by glennmatthews**
# What's Changed

- Added `--parallel` flag to `invoke unittest` in CI.
- Fixed code coverage calculation when running `invoke unittest --parallel`.
- Changed `invoke unittest` and `invoke integration-test` to automatically report code coverage on successful completion. 

# Screenshots

Running just `nautobot.core.tests` in parallel gives us 68% coverage, FWIW.

```
❯ invoke unittest --cache-test-fixtures --parallel --skip-docs-build --no-buffer --keepdb --label nautobot.core.tests
Running docker compose command "ps --services --filter status=running"
Running docker compose command "run --rm --entrypoint 'coverage erase' nautobot"
Running docker compose command "ps --services --filter status=running"
Running docker compose command "run --rm --entrypoint 'coverage run --parallel-mode --rcfile=.coveragerc.parallel 
--module nautobot.core.cli test nautobot.core.tests --config=nautobot/core/tests/nautobot_config.py 
--cache-test-fixtures --keepdb --parallel' nautobot"
Using NautobotPerformanceTestRunner to run tests ...
Using existing test database for alias 'default'...
Flushing all existing data from the database "default"...
Loading factory data from file development/factory_dump.json
Installed 2900 object(s) from 1 fixture(s)
Database default populated successfully!
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
System check identified some issues:

WARNINGS:
<Settings "nautobot_config">: (nautobot.core.W005) STORAGE_CONFIG has been set but STORAGE_BACKEND is not defined. STORAGE_CONFIG will be ignored.

System check identified 1 issue (0 silenced).
.......................................................ss.........................................................................................................................................................................................................s..........s..................................................................................s.........................................
........



......
..................................s.......
----------------------------------------------------------------------
Ran 450 tests in 124.887s

OK (skipped=6)
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Destroying test database for alias 'default'...
Emptying test database "default"...
Database default emptied!
Preserving test database for alias 'default'...
Running docker compose command "ps --services --filter status=running"
Running docker compose command "run --rm --entrypoint 'coverage combine' nautobot"
Combined data file .coverage.207886cc7164.50.627112
Combined data file .coverage.207886cc7164.47.735570
Combined data file .coverage.207886cc7164.51.988505
Combined data file .coverage.207886cc7164.49.785006
Combined data file .coverage.207886cc7164.48.378372
Combined data file .coverage.207886cc7164.46.722859
Combined data file .coverage.207886cc7164.1.757244
Running docker compose command "ps --services --filter status=running"
Running docker compose command "run --rm --entrypoint 'coverage report --skip-covered --include 'nautobot/*'' nautobot"
Name                                                                                                     Stmts   Miss  Cover
----------------------------------------------------------------------------------------------------------------------------
nautobot/__init__.py                                                                                        14      8    43%
nautobot/circuits/api/views.py                                                                              30      1    97%
(...snip....)
nautobot/virtualization/views.py                                                                           245     73    70%
----------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                    42239  13706    68%

378 files skipped due to complete coverage.
```

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- n/a Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by glennmatthews** [comment]
CI passed on rerun but there was a failure on the first run of the Postgres tests that looks a bit concerning:

```
======================================================================
FAIL: test_list_objects_csv (nautobot.dcim.tests.test_api.LocationTest)
GET a list of objects in CSV format as an authenticated user with permission to view some objects.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/runner/.cache/pypoetry/virtualenvs/nautobot-zO_ZO7c3-py3.8/lib/python3.8/site-packages/django/test/utils.py", line 387, in inner
    return func(*args, **kwargs)
  File "/home/runner/work/nautobot/nautobot/nautobot/core/testing/api.py", line 592, in test_list_objects_csv
    self.assertIn(f"cf_{cf.key}", rows[0])
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/unittest/case.py", line 1179, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: 'cf_multiple_choice' not found in {'name': 'Campus-01', 'display': 'Campus-01', 'id': 'aab49d27-8667-464b-9892-4647ffa5a664', 'object_type': 'dcim.location', 'natural_slug': 'campus-01_aab4', 'tree_depth': '0', 'time_zone': 'NULL', 'circuit_count': '8', 'device_count': '1', 'prefix_count': '0', 'rack_count': '0', 'virtual_machine_count': '0', 'vlan_count': '0', 'description': 'Edge Congress most general sell education.', 'facility': '', 'asn': '65001', 'physical_address': '', 'shipping_address': '', 'latitude': 'NULL', 'longitude': '109.040000', 'contact_name': '', 'contact_phone': '001-662-552-8112x60999', 'contact_email': 'peggy56@example.com', 'comments': '', 'parent__name': 'NoObject', 'parent__parent__name': 'NoObject', 'parent__parent__parent__name': 'NoObject', 'parent__parent__parent__parent__name': 'NoObject', 'parent__parent__parent__parent__parent__name': 'NoObject', 'location_type__name': 'Campus', 'status__name': 'Active', 'tenant__name': 'Smith-Lutz', 'created': '2024-01-23T19:41:24.738149Z', 'last_updated': '2024-01-23T19:41:24.761288Z', 'tags': 'Blue,Dark green,Dark red,Light blue,Pink,Teal,Yellow', 'cf_example_plugin_auto_custom_field': 'Default value'}
```

The `Multiple Choice` custom field is created for Locations in `nautobot.core.tests.test_views.FilterFormsTestCase.test_filtering_on_custom_select_filter_field` and so it *shouldn't* be present in `nautobot.dcim.tests.test_api.LocationTest.test_list_objects_csv`, but looks like we hit a case where the lru_cache perhaps leaked between test processes or something?

**Comment by glennmatthews** [comment]
Another timing/data-leakage failure, this time in MySQL:

```
======================================================================
FAIL: test_list_objects_csv (nautobot.dcim.tests.test_api.LocationTest.test_list_objects_csv)
GET a list of objects in CSV format as an authenticated user with permission to view some objects.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/home/runner/.cache/pypoetry/virtualenvs/nautobot-zO_ZO7c3-py3.11/lib/python3.11/site-packages/django/test/utils.py", line 387, in inner
    return func(*args, **kwargs)
    ^^^^^^^^^^^^^^^^^
  File "/home/runner/work/nautobot/nautobot/nautobot/core/testing/api.py", line 592, in test_list_objects_csv
    self.assertIn(f"cf_{cf.key}", rows[0])
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/unittest/case.py", line 1140, in assertIn
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/unittest/case.py", line 703, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: 'cf_choice' not found in {'name': 'Campus-01', 'display': 'Campus-01', 'id': 'aab49d27-8667-464b-9892-4647ffa5a664', 'object_type': 'dcim.location', 'natural_slug': 'campus-01_aab4', 'tree_depth': '0', 'time_zone': 'NULL', 'circuit_count': '7', 'device_count': '0', 'prefix_count': '1', 'rack_count': '0', 'virtual_machine_count': '0', 'vlan_count': '0', 'description': 'Edge Congress most general sell education.', 'facility': '', 'asn': '65001', 'physical_address': '', 'shipping_address': '', 'latitude': 'NULL', 'longitude': '109.040000', 'contact_name': '', 'contact_phone': '001-662-552-8112x60999', 'contact_email': 'peggy56@example.com', 'comments': '', 'parent__name': 'NoObject', 'parent__parent__name': 'NoObject', 'parent__parent__parent__name': 'NoObject', 'parent__parent__parent__parent__name': 'NoObject', 'parent__parent__parent__parent__parent__name': 'NoObject', 'location_type__name': 'Campus', 'status__name': 'Active', 'tenant__name': 'Patterson Inc', 'created': '2024-01-23T21:03:48.078217Z', 'last_updated': '2024-01-23T21:03:48.108104Z', 'tags': 'Blue,Dark green,Dark red,Light blue,Pink,Teal,Yellow', 'cf_example_plugin_auto_custom_field': 'Default value'}
```

File: nautobot/docs/development/core/getting-started.md [hunk 1]
```diff
@@ -632,20 +632,40 @@ Unit tests are run using the `invoke unittest` command (if using the Docker deve
...
-!!! warning
-    In some cases when tests fail and exit uncleanly it may leave the test database in an inconsistent state. If you encounter errors about missing objects, remove `--keepdb` and run the tests again.
+##### Unit Test Invocation Examples
 
-+/- 1.5.11
-    The `--cache-test-fixtures` argument was added to the `invoke unittest` and `nautobot-server test` commands to allow for caching of test factory data between test runs. See the [factories documentation](./testing.md#factory-caching) for more information.
+In general, when you first run the Nautobot tests in your local copy of the repository, we'd recommend:
+
+```no-highlight
+invoke unittest --cache-test-fixtures --keepdb --parallel
```
Comment (glennmatthews) [inline nautobot/docs/development/core/getting-started.md:653]: Long term I'd love to see us to make `cache-test-fixtures` and `parallel` be on by default as opt-out parameters instead.

File: nautobot/extras/tests/test_datasources.py [hunk 2]
```diff
@@ -122,6 +122,7 @@ def tearDown(self):
             del sys.modules[f"{self.repo.slug}.jobs"]
         if f"{self.repo.slug}" in sys.modules:
             del sys.modules[f"{self.repo.slug}"]
+        super().tearDown()
```
Comment (HanlinMiao) [inline nautobot/extras/tests/test_datasources.py:125]: Why is this needed?

File: nautobot/core/testing/mixins.py [hunk 3]
```diff
@@ -62,6 +62,13 @@ def setUpNautobot(self, client=True, populate_status=False):
             # Force login explicitly with the first-available backend
             self.client.force_login(self.user)
 
+    def tearDown(self):
+        """Clear lru_cache data to avoid leakage of information between test cases when running in parallel."""
```
Comment (HanlinMiao) [inline nautobot/core/testing/mixins.py:66]: Are these the only models that use LRU cache?

File: nautobot/extras/tests/test_datasources.py (hunk 2 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_datasources.py:125]: General good practice now that we have a `tearDown()` in the base class - it shouldn't be strictly necessary in this specific case at this specific time but will protect us in the eventuality that we add more to the base class method. As far as I could tell with some quick greps, this was the only test case we have with a `tearDown()` of its own that wasn't already calling super.

File: nautobot/core/testing/mixins.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/core/testing/mixins.py:66]: AFAIK yes. There's a few other uses of lru_cache in some authentication related code but those don't seem to be impacting tests.
